### PR TITLE
Implementation of PEP 673 (typing.Self)

### PR DESCRIPTION
**Pull request by Gobot1234**
### Description

An initial implementation of typing.Self type checking for [PEP 673](https://www.python.org/dev/peps/pep-0673). 

Couple of questions:
- Is this implementation actually good?
- Should think there should be a note about using the old style of this if it's detected?

## Test Plan

I've tested the use cases from the PEP individually and it worked but unit test wise I'm not entirely sure what to do should I basically just copy all of https://github.com/python/mypy/blob/master/test-data/unit/check-selftype.test and add some extra stuff for recursive classes and staticmethods?

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Did you do this off an old checkout? There's a ton of merge conflicts.

The PEP itself suggested implementing Self by desugaring into a bound TypeVar, which might be easier since it can (hopefully) be done fully in the semanal phase.

**Comment by Gobot1234** [comment]
> Did you do this off an old checkout? There's a ton of merge conflicts.
> 
> The PEP itself suggested implementing Self by desugaring into a bound TypeVar, which might be easier since it can (hopefully) be done fully in the semanal phase.

1. Yeah this is quite old, I'll fix the merge conflicts.
2. I know the PEP suggested desugaring into a bound TypeVar but I really couldn't make that work here. Something was calling deserialize and things were being replaced with FakeInfo which the only work around I found for was replacing the instance value with an unserialized one.

File: mypy/types.py [hunk 1]
```diff
@@ -447,6 +447,31 @@ def deserialize(cls, data: JsonDict) -> 'TypeVarType':
         )
 
 
+class SelfType(ProperType):
+    def __init__(self, instance: 'Instance', fullname: str, line: int = -1, column: int = -1) -> None:
```
Comment (sobolevn) [inline mypy/types.py:451]: We now also require `__slots__` for all classes in `types.py`

File: mypy/types.py [hunk 2]
```diff
@@ -447,6 +447,31 @@ def deserialize(cls, data: JsonDict) -> 'TypeVarType':
...
+        super().__init__(line, column)
+        self.name = "Self"
+        self.fullname = fullname
+        self.instance: 'Instance' = instance
```
Comment (sobolevn) [inline mypy/types.py:455]: The annotation here is probably not needed 🤔

File: mypy/types.py [hunk 3]
```diff
@@ -447,6 +447,31 @@ def deserialize(cls, data: JsonDict) -> 'TypeVarType':
         )
 
 
+class SelfType(ProperType):
+    def __init__(self, instance: 'Instance', fullname: str, line: int = -1, column: int = -1) -> None:
+        super().__init__(line, column)
+        self.name = "Self"
```
Comment (sobolevn) [inline mypy/types.py:453]: Is this going to be changed? If not, it looks like a good candidate for a class-level value.

File: mypy/type_visitor.py [hunk 4]
```diff
@@ -64,6 +64,9 @@ def visit_type_var(self, t: TypeVarType) -> T:
         pass
 
     @abstractmethod
+    def visit_self_type(self, t: SelfType) -> T:
+      pass
```
Comment (sobolevn) [inline mypy/type_visitor.py:68]:
```suggestion
        pass
```

File: mypy/type_visitor.py [hunk 5]
```diff
@@ -23,13 +23,13 @@
...
     UnboundType, ErasedType, StarType, EllipsisType, TypeList, CallableArgument,
-    PlaceholderType, TypeAliasType, ParamSpecType, get_proper_type
+    PlaceholderType, TypeAliasType, ParamSpecType, SelfType, get_proper_type
 )
 
 
 @trait
 @mypyc_attr(allow_interpreted_subclasses=True)
-class TypeVisitor(Generic[T]):
+class TypeVisitor(Generic[T], metaclass=ABCMeta):
```
Comment (sobolevn) [inline mypy/type_visitor.py:32]: Why do we need an extra metaclass here? 🤔

**Review by sobolevn — COMMENTED** [review]
Awesome, very excited for this feature! 👍

File: mypy/type_visitor.py (hunk 5 above)
Comment (Gobot1234) [reply mypy/type_visitor.py:32]: I just had it in there to catch instantiations of subclasses where I hadn't implemented visit_self_type

File: mypy/types.py (hunk 1 above)
Comment (Gobot1234) [reply mypy/types.py:451]: Will do

**Comment by MathiasSven** [comment]
I am not completely sure it is due to the implementation, but in the piece of code bellow mypy cannot correctly understand the type of an object that uses this commit's implementation of `Self`.
```python
from typing import Generic, TypeVar, Any, Optional, Protocol, Callable
from typing_extensions import Self
from abc import ABC

T = TypeVar("T")
K = TypeVar("K")

class ClassProperty(Protocol[T, K]):
    def __init__(self, getter: Callable[[T], K]) -> None: ...
    def __get__(self, obj: None, type: type[T] | None = ...) -> K: ...

def classproperty(f: Callable[[T], K]) -> ClassProperty[T, K]: ...

class ItemSet(Generic[T]):
    def first(self) -> T: ...

class BaseItem(ABC):
    @classproperty
    def set(cls) -> ItemSet[Self]: ...

class FooItem(BaseItem):
    name: str
    def test(self) -> None: ...

reveal_type(FooItem.set)
reveal_type(FooItem.set.first().test())
```
`mypy 0.920+dev.f844fbe435719f8fd756573fcc1e95ac3a3b18cd`:
```python
t.py:8: error: Invariant type variable "T" used in protocol where contravariant one is expected  [misc]
t.py:8: error: Invariant type variable "K" used in protocol where covariant one is expected  [misc]
t.py:25: note: Revealed type is "t.ItemSet*[Self@t.BaseItem]"
t.py:26: error: "object" has no attribute "test"  [attr-defined]
t.py:26: note: Revealed type is "Any"
```
`pyright 1.1.197`:
```python
t.py:25:13 - info: Type of "FooItem.set" is "ItemSet[FooItem]"
t.py:26:13 - info: Type of "FooItem.set.first().test()" is "None"
```
The contravariant and covariant issue is separate and I believe it is a problem with master too, by following the prompt it just leads to pyright complaining that T and K should be invariant and then for both it leads to "Cannot use a contravariant type variable as return type", so I assume they should be invariant.

**Comment by Gobot1234** [comment]
I'd like a bit of advice on how to be substitute Self into a function signature, should I use a TypeVistor subclass or should I just have a function to do it?

File: mypy/typeshed/stdlib/typing.pyi [hunk 6]
```diff
@@ -84,6 +84,10 @@ if sys.version_info >= (3, 10):
     TypeAlias: _SpecialForm = ...
     TypeGuard: _SpecialForm = ...
 
+if sys.version_info >= (3, 11):
+    # Self is also a (non-subscriptable) special form.
+    Self: object = ...
```
Comment (JelleZijlstra) [inline mypy/typeshed/stdlib/typing.pyi:89]: We shouldn't change this directly, it is just copied from typeshed

File: mypy/typeanal.py [hunk 7]
```diff
@@ -374,6 +374,12 @@ def try_analyze_special_unbound_type(self, t: UnboundType, fullname: str) -> Opt
                 self.fail("NotRequired[] must have exactly one type argument", t)
                 return AnyType(TypeOfAny.from_error)
             return RequiredType(self.anal_type(t.args[0]), required=False)
+        elif fullname in ('typing_extensions.Self', 'typing.Self'):
+            try:
+                bound = self.named_type(self.api.type.fullname)  # type: ignore
```
Comment (JelleZijlstra) [inline mypy/typeanal.py:379]: Should do the appropriate `isinstance` checks instead, that's cleaner and probably makes mypyc happier

**Comment by cdce8p** [comment]
Any update @Gobot1234? It would be awesome if mypy could support `Self`.

**Comment by Gobot1234** [comment]
> Any update @Gobot1234? It would be awesome if mypy could support `Self`.

I think there's a bug in the implementation but I can't find it and I don't know if I'll have time to find it with exams coming up

**Comment by Gobot1234** [comment]
That'd be fine by me, maybe they could also have a go at trying to make it equivalent to a bound TypeVar

**Comment by rattrayalex** [comment]
I'd be happy to sponsor work to complete this if that'd be helpful; just reach out.

**Comment by erikkemperman** [comment]
I've been playing with this a little bit, managed to merge everything from the master branch and fixed a few issues. So I feel like I'm making decent progress, but while adding in some unit tests I am not completely sure what to do with `__new__`.

On the one hand, the PEP explicitly disallows `Self` in static methods. But it does mention annotating the return type of `__new__` as `Type[Self]`. It says nothing about annotating the first argument. I guess we want to allow this, for regular classes? I say this because currently, if you don't annotate the first argument, mypy will reveal it as Type[<enclosing class>].

But the strange thing is, at least to me, that if I explicitly decorate `__new__` as `staticmethod`, mypy handles it differently.

```python3
class foo:
    # @staticmethod is implied
    def __new__(cls, *args, **kwargs) -> 'foo':
        reveal_type(cls)  # N:Revealed type is "Type[__main__.foo]"
        return super().__new__(cls, *args, **kwargs)


class bar:
    @staticmethod
    def __new__(cls, *args, **kwargs) -> 'bar':
        reveal_type(cls)  # N: Revealed type is "Any"
        return super().__new__(cls, *args, **kwargs)
```

Shouldn't that decoration be completely ignored? That's what I expected, anyway, since the language says that it's implicitly static but doesn't, as far as I understand, really disallow marking it static anyway.

Another issue is `__new__` in metaclasses. The PEP disallows `Self` in metaclasses outright, but currently mypy does have some special handling of the first argument. What do we want do there?

**Comment by NeilGirdhar** [comment]
> mypy will reveal it as Type[<enclosing class>]

I'm not sure, but it seems that the [documentation](https://docs.python.org/3/reference/datamodel.html#object.__new__) says nothing about `cls` being restricted to the type of (or even a subclass of) the enclosing class.  Couldn't it be any class?

>  But it does mention annotating the return type of __new__ as Type[Self]

That's also weird because the documentation doesn't promise that either: "The return value of `__new__()` should be the new object instance (**usually** an instance of cls)."

Perhaps things have changed since the docs were written, but it seems that all you can promise a priori is:
```python
def __new__(cls: type, ...) -> Any
```
It's not even guaranteed that the return value has the same type.

**Comment by thomkeh** [comment]
For what it's worth, pyright seems to allow `Self` in `__new__` in meta classes:

https://github.com/microsoft/pyright/blob/d2d97928fe5e4aaa7a5f275174d36f77e3b50e10/packages/pyright-internal/src/tests/samples/metaclass9.py#L7

**Comment by Gobot1234** [comment]
Hi @erikkemperman now that my exams are over would you be ok with me co-authoring you on this as I try to push it over the finish line?

**Comment by erikkemperman** [comment]
@Gobot1234 That’d be wonderful, actually — I got a bit stuck to be honest, and have been too busy with work stuff to work on this.

For what it’s worth, here is my branch:

https://github.com/erikkemperman/mypy/commits/gobot-master

This is basically your PR, merged with the upstream master branch of end of May, plus some some small fixes to make some tests pass.

At that point I had the feeling that it should be possible to implement this PEP in a somewhat “smaller” approach, i.e. the desugaring strategy you also mentioned, but I did not get very far trying that for lack of time.

I also was slightly confused about some edge cases like `__new__` as I mentioned above, but I guess we ought to just follow the signature currently given in typeshed.

I still think it’s a bit odd that mypy handles this method differently if a user explicitly decorates it as static, but that should probably be a separate issue.

**Comment by NeilGirdhar** [comment]
Looks like the interaction with class methods is messed up.  When `Self` is returned from a class method, it should mean the an instantiated object right?

**Comment by Gobot1234** [comment]
# steam.py
> ```diff
> + steam/enums.py:110: error: Incompatible types in assignment (expression has type "object", variable has type "Enum")  [assignment]
> + steam/enums.py:111: error: "None" has no attribute "value"  [attr-defined]
> ```
I'm able to fix these locally I can push these changes at some point later in the week

> ```diff
> + steam/enums.py:237: error: Incompatible return value type (got "object", expected "object")  [return-value]
> + steam/enums.py:241: error: Incompatible return value type (got "object", expected "object")  [return-value]
> + steam/enums.py:245: error: Incompatible return value type (got "object", expected "object")  [return-value]
> ```
These are concerning but I'm not entirely what's causing them, `cls.__new__` reveals the correct type (it's probably related to the other errors in discord.py)

> ```diff
> + steam/channel.py:185: error: Type argument "object" of "GroupChannelHistoryIterator" must be a subtype of "Union[ClanChannel, GroupChannel]"  [type-var]
> + steam/channel.py:186: error: Incompatible return value type (got "GroupChannelHistoryIterator[GroupM_co, <nothing>]", expected "GroupChannelHistoryIterator[GroupM_co, object]")  [return-value]
> ```
First error is catching a valid type error not sure what the second is on about

> ```diff
> + steam/clan.py:277: error: Incompatible types in assignment (expression has type "Union[User, SteamID]", variable has type "Optional[User]")  [assignment]
> ```
Is a type error on my end.

> ```diff
> + steam/clan.py:442: error: Incompatible return value type (got "Clan", expected "Event")  [return-value]
> + steam/clan.py:454: error: Incompatible return value type (got "Clan", expected "Announcement")  [return-value]
> ```
This error is a genieuly baffling one I'm yet to wrap my head round. These functions (`fetch_event` and `fetch_announcement`) return `Event` and `Announcement`s respectively (they don't even use `Self`) and this type checks correctly with pyright. 
The only thing that I think could be a cause for this is that all of `Clan`, `Event` and `Announcement` share a base class that performs some async initialisation and then returns the object in place (using `Self`) and that binding is somehow being overwritten but I can't reproduce this with a snippet.

> ```diff
> + steam/ext/commands/commands.py:861: error: Incompatible return value type (got "List[Command[Any]]", expected "List[Group[Any]]")  [return-value]
> ``` 
It seems like this case needs a bit of special casing for the compatiblity checker.

# discord.py
> ```diff
> + discord/flags.py:131: error: Signatures of "__ior__" and "__or__" are incompatible
> + discord/flags.py:135: error: Signatures of "__iand__" and "__and__" are incompatible
> + discord/flags.py:139: error: Signatures of "__ixor__" and "__xor__" are incompatible
> ```
Something very funky is going on here considering (for or) they are
```py
test.py:212: error: Signatures of "__ior__" and "__or__" are incompatible
test.py:215: note: Revealed type is "def (__main__.Flags, Self@__main__.Flags) -> Self@__main__.Flags"
test.py:216: note: Revealed type is "def (__main__.Flags, Self@__main__.Flags) -> Self@__main__.Flags"
```

> ```diff
> - discord/flags.py:1470: error: No overload variant of "__new__" of "type" matches argument type "Type[Self?]"
> + discord/flags.py:1470: error: No overload variant of "__new__" of "type" matches argument type "Type[object]"
> - discord/flags.py:1556: error: "Type[Self?]" has no attribute "VALID_FLAGS"
> + discord/flags.py:1556: error: "Type[object]" has no attribute "VALID_FLAGS"
> ```
Something here isn't being bound correctly.

**Comment by Gobot1234** [comment]
For the `+ steam/iterators.py:374: error: Argument 1 to "append" of "list" has incompatible type "Type[TradeOffer]"; expected "TradeOffer"  [arg-type]` I think the error is due TradeOffer being defined in another file which is weird.
```py
class TradeOffer:
    @classmethod
    def _from_history(cls) -> Self:
        trade = cls()
        return trade

class AsyncIterator(Generic[T]):
    _fill: AsyncGeneratorType[T, None]
    def __aiter__(self) -> Self:
        return self

    def __anext__(self) -> Coroutine[None, None, T]:
        return self._fill.__anext__()

    async def next(self) -> T:
        return await self.__anext__()

    async def fill(self) -> AsyncGenerator[T, None]:
        raise NotImplementedError
        yield

class TradesIterator(AsyncIterator["TradeOffer"]):
    async def fill(self) -> AsyncGenerator[TradeOffer, None]:
        async def get_trades(page: int = 100) -> list[TradeOffer]:
            trades: list[TradeOffer] = []
            for trade in ():
                trades.append(TradeOffer._from_history())
            return trades
        for trade in await get_trades():
            yield trade
```
works.

I also haven't been able to reproduce the tornado errors with a small snippet
```py
import unittest

class AsyncTestCase(unittest.TestCase):
    ...

class AsyncHTTPTestCase(AsyncTestCase):
    ...

class WebSocketBaseTestCase(AsyncHTTPTestCase):
    @types.coroutine
    def foo(self, path: str, **kwargs: Any):
        yield

@overload
def gen_test(
    *, timeout: Union[float, None] = None
) -> Callable[[Callable[..., Union[Generator, "Coroutine"]]], Callable[..., None]]:
    pass


@overload  # noqa: F811
def gen_test(func: Callable[..., Union[Generator, "Coroutine"]]) -> Callable[..., None]:
    pass


def gen_test(  # noqa: F811
    func = None,
    timeout=None,
):
    pass


class WSTest(WebSocketBaseTestCase):
    @gen_test
    def bar(self: Any):
        yield reveal_type(self.assertEqual())  # revealed type is Any
```

**Comment by ilevkivskyi** [comment]
Superseded by https://github.com/python/mypy/pull/14041

**Pull request closed without merging**